plugins/modules/foreman_hostgroup.py

```diff
@@ -267,133 +270,53 @@ def main():
-
-        for proxy in ['puppet_proxy', 'puppet_ca_proxy', 'openscap_proxy', 'content_source']:
-            if proxy in entity_dict:
-                entity_dict[proxy] = module.find_resource_by_name('smart_proxies', entity_dict[proxy], thin=True)
-
         if 'organization' in entity_dict:
             if 'organizations' in entity_dict:
                 if entity_dict['organization'] not in entity_dict['organizations']:
                     entity_dict['organizations'].append(entity_dict['organization'])
             else:
                 entity_dict['organizations'] = [entity_dict['organization']]
-            entity_dict['organization'] = module.find_resource_by_name('organizations', name=entity_dict['organization'], failsafe=False, thin=True)
-            scope = {'organization_id': entity_dict['organization']['id']}
-
-        if 'lifecycle_environment' in entity_dict:
-            entity_dict['lifecycle_environment'] = module.find_resource_by_name('lifecycle_environments', name=entity_dict['lifecycle_environment'],
-                                                                                params=scope, failsafe=False, thin=True)
-
-        if 'kickstart_repository' in entity_dict:
-            entity_dict['kickstart_repository'] = module.find_resource_by_name('repositories', name=entity_dict['kickstart_repository'],
-                                                                               params=scope, failsafe=False, thin=True)
-
-        if 'content_view' in entity_dict:
-            entity_dict['content_view'] = module.find_resource_by_name('content_views', name=entity_dict['content_view'],
-                                                                       params=scope, failsafe=False, thin=True)
 
-    entity_dict = module.handle_taxonomy_params(entity_dict)
-
-    entity = module.find_resource_by_title('hostgroups', title=build_fqn(name, parent), failsafe=True)
-    if entity:
-        entity['root_pass'] = None
-        if 'updated_name' in entity_dict:
-            entity_dict['name'] = entity_dict.pop('updated_name')
-
-    parameters = entity_dict.get('parameters')
+    module.connect()
+    entity, entity_dict = module.resolve_entities(entity_dict)
 
     puppetclasses = entity_dict.pop('puppetclasses', None)
     current_puppetclasses = entity.pop('puppetclasses', None) if entity else []
 
-    hostgroup = module.ensure_entity('hostgroups', entity_dict, entity)
+    # Remove manualy organization from entity_spec if present because it is just used for katello's scoped resources
+    entity_dict.pop('organization', None)
 
-    if hostgroup:
-        scope = {'hostgroup_id': hostgroup['id']}
-        module.ensure_scoped_parameters(scope, entity, parameters)
+    entity = module.ensure_entity('hostgroups', entity_dict, entity)
+    module.ensure_entity_parameters(entity, entity_dict)
```

Comment:
check_mode is failing with

```
"msg": "Failed to list resource: Can't overwrite existing cassette ('fixtures/hostgroup-0.yml') in your current record mode ('none').\nNo match for the request (<Request (GET) https://foreman.example.com/api/hostgroups/-1/parameters?per_page=4294967296>) was found.\nFound 1 similar requests with 1 different matcher(s) :\n\n1 - (<Request (GET) https://foreman.example.com/api/hostgroups/5/parameters?per_page=4294967296>).\nMatchers succeeded : ['method', 'query_matcher_ignore_proxy', 'body_json_l2_matcher']\nMatchers failed :\npath - assertion failure :\n/api/hostgroups/-1/parameters != /api/hostgroups/5/parameters\n"
```

this is because you call `ensure_entity_parameters` for a "just" created hostgroup, but in check mode you didn't really create it, so the `id` you got back is `-1`.

Reply:
but I also might be wrong here and @mdellweg needs to help :D

Reply:
I can only speculate what `ensure_entity_parameters` is for, but your analysis sounds reasonable.

Reply:
I think, in `ensure_entity_parameters` there should be a safeguard to not search for existing parameters if we already know, we _just_ created the resource.